### PR TITLE
ARROW-12743 [R] Add DESCRIPTION fields for dev dependencies

### PR DESCRIPTION
**Pull request by dragosmg**
This PR makes it easier to install additional dependencies by using the `Config/Needs/...` fields in DESCRIPTION.

```
remotes::install_deps(dependencies = c("Config/Needs/dev", "Config/Needs/website"))
```

### CODE REVIEW

**Comment by dragosmg** [comment]
Bumped into this by accident, after installing the dev version of {roxygen2} with the above command.
image

**Comment by jonkeane** [comment]
This seems fine, would you mind also changing / adding to the developer docs to reflect what this change would mean? Same for the make file: could you add in what would need to change there?

You also should look through the repo and see if there are other places (i.e. in our CI) that could benefit from this (presumably, one of the benefits is that we can control what versions / where from those are installed from one place now rather than needing to do it in a bunch of different places?)

**Comment by dragosmg** [comment]
Not sure what to recommend in the developer docs. I ran 
```
install_deps(".", dependencies = c("soft", "Config/Needs/1", "Config/Needs/2"))
```
and 40+ minutes later my R session was still busy installing (mostly binary) packages. `dependencies = "soft"` is the equivalent of `dependencies = TRUE` but something is not right with this picture.

I'd be tempted to suggest using pak as that uses a separate process and is much faster.

**Comment by jonkeane** [comment]
> 40+ minutes later my R session was still busy installing (mostly binary) packages

Indeed, something sounds not quite right about that or not the behavior we would ultimately want here.

As for what to recommend in the developer docs I like to approach it from the questions of: Who is our audience there? What do they need to accomplish with these docs? What skills are they coming in with? (there almost certainly are multiple answers here, so we'll likely need to thread that needle between the different groups) Sometimes I find that writing out what the docs are (or what are the commands that I would run to do this task and what do those look like? Here those _are_ the docs cause we're not writing a new function here, but the principle is the same) helps figure out if the UX feels right or if something is missing there

**Comment by dragosmg** [comment]
@thisisnic & @jonkeane would you mind having a look? 🙏🏻

**Review by jonkeane — COMMENTED** [review]
This is looking good. 

Do we also want to edit https://github.com/apache/arrow/blob/master/ci/scripts/r_deps.sh and use this new way of installing? 

That script is called by many of our CI jobs (and many of those don't need things like pkgdown, etc), but we _do_ install pkgdown manually now in the linux-apt-docs.docker file: https://github.com/apache/arrow/blob/69c25ed136da984718e7fbb5de83ee11515809e8/ci/docker/linux-apt-docs.dockerfile#L94

Which we should be able to do something like `RUN /arrow/ci/scripts/r_deps.sh /arrow ...`

And have the argument passed down to `r_deps.sh` (or this could be accomplished with an environment variable if you preferred.

You probably already know about his, but just in case, there are some great docs about dunning these locally that can be helpful in debugging: https://ursalabs.org/arrow-r-nightly/articles/developers/docker.html#running-existing-workflows-from-docker-compose-yml

**Comment by dragosmg** [comment]
There is a [new GHA workflow for installing dependencies](https://github.com/r-lib/actions/tree/v2-branch/setup-r-dependencies). Do we want to make use of it or should I aim to close this issue and then open a follow-up Jira for later? It looks like it allows us to control how `Config/Needs` are installed, via:
```
needs: |
      website
      coverage
```

**Comment by thisisnic** [comment]
> There is a [new(ish) GHA workflow for installing dependencies](https://github.com/r-lib/actions/tree/v2-branch/setup-r-dependencies). Do we want to make use of it or should I aim to close this issue and then open a follow-up Jira for later? It looks like it allows us to control how `Config/Needs` are installed, via:
> 
> ```
> needs: |
>       website
>       coverage
> ```

I'd be inclined to look at it in a follow-up ticket, to keep things a bit simpler on the ticket and PR discussion

**Comment by dragosmg** [comment]
@github-actions crossbow submit test-ubuntu-default-docs

**Comment by dragosmg** [comment]
@github-actions crossbow submit test-ubuntu-default-docs

**Comment by dragosmg** [comment]
I think I am on the right track as the documentation job is failing with a code 3 which indicates we are on the correct branch of the conditional statement in `r_deps.sh`
```
elif [ ${R_DEV_DEPS} = "website" ]; then
  exit 3
  ${R_BIN} -e "remotes::install_deps(dependencies = 'Config/Needs/website')"
  ```

**Comment by dragosmg** [comment]
@github-actions crossbow submit test-ubuntu-default-docs

**Comment by dragosmg** [comment]
I think the docs might be failing due to `remotes::install_deps(dependencies = c("Config/Needs/webiste"))` only installing `pkgdown` and not it's dependencies, one of which is `fs`.

File: r/vignettes/developers/workflow.Rmd [hunk 1]
```diff
@@ -4,6 +4,22 @@
 knitr::opts_chunk$set(error = TRUE, eval = FALSE)
 ```
 
+The Arrow R package uses several additional development tools:
+
+* `lintr` for code analysis
+  - for the time being, the R package uses a custom version of lintr - `jonkeane/lintr@arrow-branch`
+* `styler` for code styling
+* `pkgdown` for building the website
+* `roxygen2` for documenting the package
```
Comment (thisisnic) [inline r/vignettes/developers/workflow.Rmd:13]: I really like how you've mentioned the packages up-front here; how about also adding links to their pkgdown sites when you click on their names?

File: r/vignettes/developers/workflow.Rmd [hunk 2]
```diff
@@ -34,7 +44,7 @@ pkgdown::build_site(preview=TRUE)
 
 The R code in the package follows [the tidyverse style](https://style.tidyverse.org/). On PR submission (and on pushes) our CI will run linting and will flag possible errors on the pull request with annotations.
 
-To run the [lintr](https://github.com/jimhester/lintr) locally, install the lintr package (note, we currently use a fork that includes fixes not yet accepted upstream, see how lintr is being installed in the file `ci/docker/linux-apt-lint.dockerfile` for the current status) and then run
+To run the [lintr](https://github.com/r-lib/lintr) locally, install the lintr package (note, we currently use a fork that includes fixes not yet accepted upstream, see how lintr is being installed in the file `ci/docker/linux-apt-lint.dockerfile` for the current status) and then run
```
Comment (thisisnic) [inline r/vignettes/developers/workflow.Rmd:47]:
```suggestion
To run the [lintr](https://github.com/r-lib/lintr) locally, install the lintr package (note, we currently use a fork that includes fixes not yet accepted upstream; see how lintr is being installed in the file `ci/docker/linux-apt-lint.dockerfile` for the current status) and then run
```

File: r/vignettes/developers/workflow.Rmd (hunk 1 above)
Comment (dragosmg) [reply r/vignettes/developers/workflow.Rmd:13]: That sounds like a great idea. We might need to that in a separate PR as I don't think we are going to merge this PR anytime soon.

Comment (dragosmg) [reply r/vignettes/developers/workflow.Rmd:13]: Done, with the exception of {lintr} which does not have a pkgdown site - I linked to the GitHub.

File: r/vignettes/developers/workflow.Rmd (hunk 2 above)
Comment (dragosmg) [reply r/vignettes/developers/workflow.Rmd:47]: Done

**Comment by dragosmg** [comment]
Given the results of the [little investigation](https://issues.apache.org/jira/browse/ARROW-15299) into the `dependencies` argument for `remotes::install_deps()` (reproduced below). What direction should we take this PR in?

Options: 
* do nothing - leave things as they are (i.e. not use the `Config/Needs/...` fields in DESCRIPTION -> close this PR without merging any of the proposed changes
* do something - cherry pick the PR and mostly keep the parts that improve a bit the documentation
* still make use of the `Config/Needs/...` fields as it allows us to keep some "workflow" dependencies closer to the code - keep the PR as is (maybe supplement calls with `dependencies = c("hard", "Config/Needs/website")` with `dependencies = TRUE`.
* do something else.

What do you think? @jonkeane @nealrichardson @thisisnic @paleolimbot @ianmcook and others 😄 .

=====================
Investigation results:
What we would need is a combination between `dependencies = TRUE` and `dependencies = "Config/Needs/website"`. I ran the experiment and the results are below:

test | description | number of packages installed
:--: | -- | :--:
1 | dependencies = TRUE | 64
2 | dependencies = 'hard' | 12
3 | dependencies = c('hard', 'Config/Needs/website') | 61
4 | dependencies = 'soft' | 1819
5 | dependencies = c('soft', 'Config/Needs/website') | 1819
6 | dependencies = c('Config/Needs/website') | 1

I excluded `dependencies = "soft"` & combinations of `"soft"` since they are _too_ recursive and end up installing an excessive number of packages.  
`dependencies = "hard"` & `dependencies = "Config/Needs/website"` are too narrow.
I compared `dependencies = TRUE` to `dependencies = c("hard", "Config/Needs/website")`. While the overall number of installed packages looks promising (64 vs 61), there isn't complete overlap.


indicator | value
-- |:--:
packages exclusive to test 1 | 21
packages exclusive to test 3 | 18
packages in common | 43

We can safely conclude that `dependencies = c("hard", "Config/Needs/website")` is **not equal** to `dependencies = TRUE` + `dependencies = c("Config/Needs/website")`.
We definitely need `"Config/Needs/website"` - in our case pkgdown and its dependencies, but in combination with `dependencies = "hard"` it is too narrow in scope, while with `dependencies = "soft"` it is much too wide.


How I got there?
* I removed all installed packages from the system library (with the exception of `"base"` and `"recommended"`packages)
* I installed remotes and rstudioapi in the system library
* each test installed packages to its own folder (configured via the R_LIBS_USER environment variable) with the R session being restarted for each run
* I then compared the contents of the 6 libraries

**Comment by dragosmg** [comment]
Discussed yesterday with @jonkeane to keep only the documentation improvement bits of this PR.

**Review by thisisnic — APPROVED** [review]
Looks good to me.  The only thing to note is that in a recent PR someone mentioned that you can refer to packages {like this} and pkgdown will automatically link to them, so if you want to change any of those links, go ahead, but it's not crucial.

**Comment by dragosmg** [comment]
Thanks. I'd be keen to merge this as-is if that's ok.

File: r/vignettes/developers/workflow.Rmd [hunk 3]
```diff
@@ -4,6 +4,23 @@
...
+* [`lintr`](https://github.com/r-lib/lintr) for code analysis
+  - for the time being, the R package uses a custom version of lintr - `jonkeane/lintr@arrow-branch`
+* [`styler`](https://styler.r-lib.org) for code styling
+* [`pkgdown`](https://pkgdown.r-lib.org) for building the website
+* [`roxygen2`](https://roxygen2.r-lib.org) for documenting the package
+  - the R documentation uses the [`@examplesIf`](https://roxygen2.r-lib.org/articles/rd.html#functions) tag introduced in `roxygen2` version 7.1.2
+
+You can install all these additional dependencies by running:
+
+```{r, eval=FALSE}
```
Comment (eitsupi) [inline r/vignettes/developers/workflow.Rmd:18]: If it is just a code block, the eval option is not needed.

```suggestion
```r
```

Comment (dragosmg) [reply r/vignettes/developers/workflow.Rmd:18]: Good catch. Thanks

**Pull request closed without merging**